src/dmd/mtype.d

```diff
@@ -4551,6 +4551,17 @@ extern (C++) final class TypeFunction : TypeNext
                 stc |= STC.return_;
             }
         }
+        else
+        {
+            // Check escaping through return value
+            Type tret = nextOf().toBasetype();
+            if (isref || tret.hasPointers())
+            {
+                return stc;
+            }
+
+            stc |= STC.scope_;
```

Comment:
I just realized there should also be a check that the function is `nothrow`, or else parameters could escape using Exceptions.
```D
import std.stdio;

void escape(string x) @safe pure {
    throw new Exception(x);
}

void main() @safe {
    static string global;
    immutable(char)[4] local;
    try {
        escape(local[]); 
    } catch(Exception e) {
        global = e.msg;
    }
    assert(&local[0] != &global[0]); // fails
}
```
Though that might be better to do in a follow-up PR, so we can get the fix for the more common case in quicker.